pkg/objectstore/bucket.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+//nolint:nonamedreturn
+package objectstore
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+	"github.com/thanos-io/objstore/client"
+	"github.com/thanos-io/objstore/providers/azure"
+	"github.com/thanos-io/objstore/providers/bos"
+	"github.com/thanos-io/objstore/providers/cos"
+	"github.com/thanos-io/objstore/providers/filesystem"
+	"github.com/thanos-io/objstore/providers/gcs"
+	"github.com/thanos-io/objstore/providers/obs"
+	"github.com/thanos-io/objstore/providers/oci"
+	"github.com/thanos-io/objstore/providers/oss"
+	"github.com/thanos-io/objstore/providers/s3"
+	"github.com/thanos-io/objstore/providers/swift"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"gopkg.in/yaml.v2"
+)
+
+func NewBucket(tracer trace.Tracer, logger log.Logger, confContentYaml []byte, reg prometheus.Registerer, component string) (objstore.InstrumentedBucket, error) {
```

Comment:
Since we're gonna want this in at least 3 repos, I'd suggest we propose this upstream.

For what it's worth I'm not a fan of `NewBucket` returning fully instrumented things, that makes assumptions that are not universally true. I think it's better for the caller to make the decision whether to instrument or not.

I would say here we just have the `NewTracingBucket` and defer the instrumentation decision to the caller.

Reply:
https://github.com/thanos-io/objstore/pull/61